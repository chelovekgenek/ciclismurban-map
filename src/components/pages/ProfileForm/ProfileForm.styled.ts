import styled from "styled-components"

export const EmailField = styled.h3``
