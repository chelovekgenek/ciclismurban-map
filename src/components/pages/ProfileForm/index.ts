export { ProfileFormContainer as ProfileForm } from "./ProfileForm.container"
