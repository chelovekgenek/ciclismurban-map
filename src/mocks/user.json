{
  "uuid": "e906f10a-dfe3-4cf4-9197-67b9de5ba91f",
  "email": "test@example.com",
  "createdAt": "2019-10-31T14:54:30.015Z",
  "updatedAt": "2019-10-31T14:54:30.015Z"
}
