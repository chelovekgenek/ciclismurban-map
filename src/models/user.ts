export { UserModel, ExposeGroup, AuthResponseModel } from "@ciclismurban/api/dist/src/modules/user/models"

export class LoginForm {
  email = ""
  password = ""
}
