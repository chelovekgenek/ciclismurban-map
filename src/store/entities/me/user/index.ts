import * as Actions from "./user.actions"
import * as Selectors from "./user.selectors"

export { Actions, Selectors }
