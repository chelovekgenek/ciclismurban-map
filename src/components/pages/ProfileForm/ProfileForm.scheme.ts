import { ProfileModel } from "@ciclismurban/models"

export class ProfileFormValues extends ProfileModel {
  constructor(props: Partial<ProfileModel> = {}) {
    super()

    if (props.avatar) {
      this.avatar = props.avatar
    }
    if (props.status) {
      this.status = props.status
    }
    if (props.description) {
      this.description = props.description
    }
    if (props.facebook) {
      this.facebook = props.facebook
    }
    if (props.telegram) {
      this.telegram = props.telegram
    }
  }
}
