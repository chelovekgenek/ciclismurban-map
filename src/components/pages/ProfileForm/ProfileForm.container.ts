import { connect } from "react-redux"
import { withFormik } from "formik"
import { get } from "lodash-es"
import { UserExposeGroup, ProfileModel } from "@ciclismurban/models"

import { TAppState } from "store/entities"
import { validateFormik } from "helpers"
import { getFetching, getProfile, MeUpdateProfileActions } from "store/entities/me"

import { IProps, ProfileForm } from "./ProfileForm"
import { ProfileFormValues } from "./ProfileForm.scheme"

export interface IStateProps {
  fetching: ReturnType<typeof getFetching>
  profile: ReturnType<typeof getProfile>
}
export interface IDispatchProps {
  save: typeof MeUpdateProfileActions.request
}

const hocConnect = connect<IStateProps, IDispatchProps, {}, TAppState>(
  state => ({
    fetching: getFetching(state),
    profile: getProfile(state),
  }),
  {
    save: MeUpdateProfileActions.request,
  },
)
const hocWithFormik = withFormik<IProps, ProfileFormValues>({
  enableReinitialize: true,
  mapPropsToValues: props => new ProfileFormValues(props.profile),
  handleSubmit: (values, { props }) => props.save(values),
  validate: values => {
    let errors = validateFormik(ProfileModel, [UserExposeGroup.UPDATE], ["avatar"])(values)
    if (!get(values, "avatar")) {
      errors.image = "avatar must be provided"
    }
    return errors
  },
  validateOnBlur: true,
})

export const ProfileFormContainer = hocConnect(hocWithFormik(ProfileForm))
