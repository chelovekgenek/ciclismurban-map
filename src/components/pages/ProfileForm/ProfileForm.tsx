import React, { useMemo } from "react"
import { Form, FormikProps } from "formik"
import { isEmpty } from "lodash-es"

import { App } from "components/generic/layout"
import { Textarea, Upload, Input } from "components/generic/form"
import { Button } from "components/generic/ui"

import { ProfileFormValues } from "./ProfileForm.scheme"
import { IDispatchProps, IStateProps } from "./ProfileForm.container"

export interface IProps extends IStateProps, IDispatchProps {}

export const ProfileForm: React.FC<IProps & FormikProps<ProfileFormValues>> = ({
  submitForm,
  resetForm,
  isValid,
  touched,
  fetching,
}) => {
  const actions = useMemo(
    () => [
      <Button key="1" onClick={() => resetForm()} disabled={isEmpty(touched)}>
        Сбросить
      </Button>,
      <Button key="2" type="primary" onClick={submitForm} disabled={!isValid} loading={fetching}>
        Сохранить
      </Button>,
    ],
    [fetching, isValid, touched, resetForm, submitForm],
  )
  return (
    <App
      content={{
        useHeader: true,
        useLayout: true,
        title: "Редактировать профиль",
        actions,
      }}
    >
      <Form>
        <Upload label="Изображение" name="avatar" />
        <Textarea label="Описание" name="description" autosize={{ minRows: 2, maxRows: 4 }} />
        <Input label="Telegram" name="telegram" />
        <Input label="Facebook" name="facebook" />
      </Form>
    </App>
  )
}
