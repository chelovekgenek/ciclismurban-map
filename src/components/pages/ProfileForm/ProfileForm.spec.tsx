import React from "react"
import { shallow } from "enzyme"

import formProps from "mocks/formik-form-props"
import user from "mocks/user.json"

import { ProfileForm } from "./ProfileForm"
import { ProfileFormValues } from "./ProfileForm.scheme"

describe("pages/ProfileForm", () => {
  const initialValues = new ProfileFormValues()
  const render = (props: Partial<React.ComponentProps<typeof ProfileForm>> = {}) =>
    shallow(
      <ProfileForm
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        fetching={false}
        profile={user.profile}
        save={jest.fn() as any}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
  it("should be able to reset form if touched", () => {
    const wrapper = render({ touched: { description: true } })
    expect((wrapper.find("App").prop("content") as any).actions[0].props.disabled).toBe(false)
  })
})
