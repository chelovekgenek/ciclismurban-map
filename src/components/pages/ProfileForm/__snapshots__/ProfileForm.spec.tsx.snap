// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/ProfileForm should match snapshot 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
        >
          Сбросить
        </Button>,
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Редактировать профиль",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Upload
      label="Изображение"
      name="avatar"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 2,
        }
      }
      label="Описание"
      name="description"
    />
    <Input
      label="Telegram"
      name="telegram"
    />
    <Input
      label="Facebook"
      name="facebook"
    />
  </Form>
</App>
`;
